### PR TITLE
consensus: save block metadata

### PR DESCRIPTION
**Pull request by pro-wh**
image

proposer and signers

### CODE REVIEW

File: analyzer/consensus/registration.go
Comment (pro-wh) [inline analyzer/consensus/registration.go:44]: this is wrong, it's not the same kind of address.

see oasis's cometbft address impl. need to make a lookup

Comment (pro-wh) [reply analyzer/consensus/registration.go]: ok after a lot of fuss, the lookup is made

File: storage/client/queries/queries.go
Comment (mitjat) [inline storage/client/queries/queries.go:39]: :hot_face: swanky. nice job.

Maybe a comment for future readers? `-- Maps signer_node_consensus_pubkey_addresses to an array of structs that each describe the corresponding node` or sth

File: go.mod
Comment (mitjat) [inline go.mod:68]: Do you understand how this was added? We're not using it directly, and the only other depedendency change is that cometbft went from being an implicit dep to an explicit one. Maybe Go only adds indirect deps when they are actually used, rather than all transitive deps?

File: api/spec/v1.yaml
Comment (mitjat) [inline api/spec/v1.yaml:1360]: I would expect to see these beauts in e2e diffs. Probably just needs to be re-run?

**Review by mitjat — COMMENTED** [review]
Thank you thank you for figuring these out!
Generally LGTM. Most of my comments are about explaining the semantics to our API consumers (and to future nexus code maintainers).

File: storage/migrations/00_consensus.up.sql
Comment (pro-wh) [reply storage/migrations/00_consensus.up.sql]: 14 now :skull:

File: analyzer/consensus/consensus.go
Comment (pro-wh) [reply analyzer/consensus/consensus.go:357]: haven't tested on damask either, back when it was a "tendermint" structure

File: analyzer/consensus/consensus.go
Comment (pro-wh) [reply analyzer/consensus/consensus.go:351]: we/our as in the oasis-core team? probably would be for consistency. different consensus backends could have different information

File: api/spec/v1.yaml
Comment (pro-wh) [reply api/spec/v1.yaml:1380]: it's like just a reference to a validator, enough stuff to render a link to their explorer page. not full details. probably does need to be named something other than "info" but imo "signer" is too specific to this use case. maybe node reference

File: api/spec/v1.yaml
Comment (pro-wh) [reply api/spec/v1.yaml:1389]: not equivalent, `type: object` is more specific. having an explicit schema would be good, and that including the "type." but I'm against being the first mover with a half measure like adding `type: object` alone. import the metadata schema throughout the codebase, and not in this pr

File: analyzer/consensus/consensus.go
Comment (mitjat) [reply analyzer/consensus/consensus.go:357]: OK let's say "pre-Eden" in the comment then; it's more informative than "earlier".

> back when it was a "tendermint" structure

And that's the structure that we "don't have", per the comment? The comment still confuses me. Is it that we cannot find that struct anymore (sounds unlikely), or that we haven't bothered importing/implementing it, or that that structure appears somewhere other than the `data.BlockHeader.Meta` field and so it does not apply here?

Comment (pro-wh) [reply analyzer/consensus/consensus.go:357]: this BlockMeta structure, we haven't been vendoring it. the comment is meant to say that we try to decode any block metadata into the latest BlockMeta structure as directly imported oasis-core, we don't have earlier version of the struct to help with decoding earlier blocks' metadata

Comment (pro-wh) [reply analyzer/consensus/consensus.go:357]: oo shouldn't return here

Comment (pro-wh) [reply analyzer/consensus/consensus.go]: comment updated

File: analyzer/consensus/consensus.go
Comment (Andrew7234) [inline analyzer/consensus/consensus.go:366]: Can we convert this to an oasis-address (assuming we just feed the bytes to a bech32 derivation) and store it like that instead? It'd be nice to avoid another address format and to keep things consistent, but it'd also be v helpful for supporting queries on signed blocks/proposed blocks/uptime for a given validator.

Same for the `prevSigners` addresses below

Also, does the `proposerAddr` address refer to the node_id (type `base64_ed25519_pubkey` in the db), but encoded in hex?

File: storage/migrations/14_block_meta.up.sql
Comment (Andrew7234) [inline storage/migrations/14_block_meta.up.sql:17]: Heads up - we have a request to support queries for proposed/signed blocks for a given validator (https://app.clickup.com/t/86941rbrt), and I imagine we'll have to add indexes on these two columns to support that (gin index for the latter). Your call on whether you want to add that in this PR though.

File: storage/migrations/14_block_meta.up.sql
Comment (ptrus) [inline storage/migrations/14_block_meta.up.sql:9]: Maybe add a comment on why these can now be null? Due to `ConsensusBlockSignersUpsert` inserting metadata for previous block, which doesn't necessarily yet exists (due to no order guarantees when processing blocks in fast-sync) right?

Comment (ptrus) [reply storage/migrations/14_block_meta.up.sql:9]: Alternatively we could have also split out the metadata into a separate table, I guess that was considered and decided against it?

Comment (pro-wh) [reply storage/migrations/14_block_meta.up.sql:9]: oh yeah could use a separate table

File: analyzer/consensus/consensus.go
Comment (pro-wh) [reply analyzer/consensus/consensus.go:362]: ok yeah

File: storage/client/client.go
Comment (mitjat) [inline storage/client/client.go:341]: Currently unused; let's not pull it from the DB in the first place. (I'm supportive of having the raw meta field in the db in case we need to extract more info from it later)

File: api/spec/v1.yaml
Comment (mitjat) [reply api/spec/v1.yaml:1366]: Ah yes, I see e2e fails with exactly the expected diffs :+1:

File: storage/client/client.go
Comment (pro-wh) [reply storage/client/client.go:341]: woa is it not used?

Comment (pro-wh) [reply storage/client/client.go:341]: huh did I intend to add that to the response or something

File: storage/migrations/14_block_meta.up.sql
Comment (mitjat) [inline storage/migrations/14_block_meta.up.sql:17]: Can we add a comment about why this is not `oasis_addr[]`, and what the actual values are here?

Alternatively, let's store the oasis addr here (as suggested in https://github.com/oasisprotocol/nexus/pull/646/files#r1591463407 by Andy). @jberci said (and Warren probably knows well) that we can derive this on the fly, rather than looking it up from the DB. The downside is that we're less robust if very hypothetically oasis-core were to change the derivation scheme; negligible risk, IMO.

File: go.mod
Comment (Andrew7234) [inline go.mod:173]: Is there a reason for this version? It seems quite old: https://pkg.go.dev/github.com/cometbft/cometbft?tab=versions

File: storage/migrations/28_block_meta.up.sql
Comment (Andrew7234) [inline storage/migrations/28_block_meta.up.sql:8]: Neat idea! Yeah fast-sync is annoying to deal with. Although if we use the node consensusAddr -> entityID mapping, we hopefully wouldn't need this or `block_signer_links` since we can store the entityID directly.

File: api/spec/v1.yaml
Comment (Andrew7234) [inline api/spec/v1.yaml:1368]: `node_id` is not needed to satisfy the consensus mocks, let's leave it out since it simplifies things. (Plus, i doubt most users would be able to use it, `entity_id` is more helpful since it corresponds to the validator)

**Review by Andrew7234 — COMMENTED** [review]
Thanks! Looking better and better, left a suggestion about tracking consensusAddr -> entity ID and some follow-up comments off of that. Let me know what you think!

File: analyzer/consensus/consensus.go
Comment (Andrew7234) [reply analyzer/consensus/consensus.go:397]: Damask worked for me, didn't try cobalt though

File: analyzer/queries/queries.go
Comment (Andrew7234) [inline analyzer/queries/queries.go:192]: Would this work if we used an `UPDATE` instead? We could then keep the `not null` constraints for the other `chain.blocks` columns. I _think_ they were originally removed to support updating the signers for the `from`-1 block (since the consensus analyzer didn't analyze the block before `from`, it won't have any of the fields populated). As a bonus, the migration becomes shorter/faster.

**Review by Andrew7234 — APPROVED** [review]
Looks good! There's a few remaining comments from ptrus/mitja which need to be addressed, but hopefully those are minor changes. Thanks

File: storage/migrations/32_block_meta.up.sql
Comment (ptrus) [inline storage/migrations/32_block_meta.up.sql:5]: What was the reasoning for storing Entity IDs instead of Node IDs here? 

A validator entity can have multiple validator nodes registered. In that case, we lose information about which exact node signed the block. 

In practice, only a single node will be active at any given epoch (since the network limits to one at this point), so this shouldn't be that problematic. However, I feel we could easily have used node IDs here to avoid this.

Comment (ptrus) [reply storage/migrations/32_block_meta.up.sql:5]: Actually, it probably makes sense/is fine, since we mostly don't care about individual validator nodes in other places/in the API.

Comment (pro-wh) [reply storage/migrations/32_block_meta.up.sql:5]: yeah, it's for reasons of API simplicity like that. the general style of nexus is to have analysis precompute stuff that the API side will need, even sometimes at the cost of lossiness

Comment (Andrew7234) [reply storage/migrations/32_block_meta.up.sql:5]: > we mostly don't care about individual validator nodes in other places/in the API.

Yeah it was a combination of this, also because we don't track historical validator entity<>node mappings (though it is possible to add now in the `history.validators` table).

File: storage/migrations/00_consensus.up.sql
Comment (ptrus) [inline storage/migrations/00_consensus.up.sql:28]: These comments are wrong  `32_block_meta.up.sql` does not make these fields `NULL`-able.

File: analyzer/consensus/consensus.go
Comment (ptrus) [inline analyzer/consensus/consensus.go:445]: This doesn't work correctly in the fast-sync case. When in fast-sync, blocks are not processed in order, so the previous block (`cmtMeta.LastCommit.Height`) might not exist, and therefore this query won't insert any singers.

Some quick suggestion on how to fix this:
- doing an Upsert, and making all other fields NULL-able (I think a previous version of this PR was done in that way)
- storing the signers in a separate table 

I noticed this in E2E tests of https://github.com/oasisprotocol/nexus/pull/764 since some blocks are missing signers in the `edenfast` test case.

**Pull request merged**
